_posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md

```diff
@@ -0,0 +1,472 @@
+that we will use (Fake IPA requires to have that list ready when it starts). A
+"node" typically looks like this
+
+```shell
+{
+      "uuid": $uuid,
+      "name": $node_name,
+      "power_state": "Off",
+      "external_notifier": "True",
+      "nics": [
+        {"mac": $macaddr, "ip": "192.168.0.100"}
+      ],
+}
+```
+
+All of the variables (`uuid`, `node_name`, `macaddress`) can be dynamically
+generated in any way you want (check [this
+script](https://github.com/Nordix/metal3-clusterapi-docs/blob/main/Support/Multitenancy/Scalability-with-Fake-Nodes/generate_unique_nodes.sh)
+out if you need an idea). Still, we must store this information to generate the
+BMH objects that match those "nodes." The `ip` is, on the other hand, not
+essential. It could be anything.
+
+We must also start up the __sushy-tools__ container in this step. It is a tool
+that simulates the [Baseboard Management
+Controller](https://www.techtarget.com/searchnetworking/definition/baseboard-management-controller)
+for non-bare-metal hardware, and we have been using it extensively inside Metal3
+dev-env and CI to control and provision VMs as if they are bare-metal nodes. In
+a bare-metal setup, Ironic will ask the BMC to install IPA on the node, and in
+our setup, __sushy-tools__ will get the same request, but it will simply fake
+the installation and, in the end, forward __Ironic__ traffic to the __Fake IPA__
+container.
+
+Another piece of information we will need is the cert that __Ironic__ will use
+in its communication with __IPA__. IPA is supposed to get it from Ironic, but as
+__Fake IPA__ cannot do that (at least not yet), we must get the cert and provide
+it in __Fake IPA__ config.
+
+```shell
+mkdir cert
+kubectl get secret -n baremetal-operator-system ironic-cert -o json \
+  -o=jsonpath="{.data.ca\.crt}" | base64 -d >cert/ironic-ca.crt
+```
+
+Also note that one set of __sushy-tools__ and __Fake IPA__ containers won't be
+enough to provision 1000 nodes. Just like __Ironic__, they need to be scaled up
+extensively (about 20-30 pairs will be sufficient for 1000 nodes), but
+fortunately, the scaling is straightforward: We just need to give them different
+ports. Both of these components also require a Python-based config file. For
+convenience, in this setup, we create a big file and provide it to both of them,
+using the following shell script:
+
+```shell
+for i in $(seq 1 "$N_SUSHY"); do
+  container_conf_dir="$SUSHY_CONF_DIR/sushy-$i"
+
+  # Use round-robin to choose fake-ipa and sushy-tools containers for the node
+  fake_ipa_port=$((9901 + (($i % ${N_FAKE_IPA:-1}))))
+  sushy_tools_port=$((8000 + i))
+  ports+=(${sushy_tools_port})
+
+  # This is only so that we have the list of the needed ports for other
+  # purposes, like configuring the firewalls.
+  ports+=(${fake_ipa_port})
+
+  mkdir -p "${container_conf_dir}"
+
+  # Use default username (`username`) and password (`password`) in base64 
```

Comment:
Sorry but this is not correct. It is an htpasswd file and the username is `admin` (first part of the string). The rest is probably a bcrypt hash of the password, which I don't know what it is. Anyway you can probably just mention that this an htpasswd file used for securing the container(s)? I'm not sure if it is used for sushy-tools or fake IPA or both though.